src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/ContainerProcessor.java

```diff
@@ -73,20 +75,43 @@ private boolean isReady(BlobItem blob, String containerName) {
     }
 
     private void processBlob(BlobClient blob) {
-        envelopeService
-            .findLastEnvelope(blob.getBlobName(), blob.getContainerName())
+        getLastEnvelope(blob)
             .ifPresentOrElse(
                 envelope -> {
                     if (envelope.status == Status.CREATED) {
-                        leaseAndThen(blob, () -> blobProcessor.continueProcessing(envelope.id, blob));
+                        leaseAndThen(blob, () -> {
+                            getLastEnvelope(blob)
+                                .ifPresentOrElse(
+                                    envelope1 -> {
+                                        if (envelope1.status == Status.CREATED) {
+                                            blobProcessor.continueProcessing(envelope.id, blob);
+                                        } else {
+                                            logProcessedEnvelope(envelope);
+                                        }
+                                    },
+                                    () -> logger.error(
+                                        "Envelope no longer exists in system for blob {}, container {}",
+                                        blob.getBlobName(), blob.getContainerName()
+                                    )
+                                );
+                        });
                     } else {
-                        logger.info("Envelope already processed in system, skipping. {} ", envelope.getBasicInfo());
+                        logProcessedEnvelope(envelope);
```

Comment:
here's a wild thought: how about we try lease straight away before making db call? acquirer already checks `exc.getErrorCode() != LEASE_ALREADY_PRESENT && exc.getErrorCode() != BLOB_NOT_FOUND` and doesn't log an error if this condition is not satisfied. which is what we need right? then on else part we'll just log info as we do now.

by the looks here we just overusing if/else/option/if/else things. plus multiple db calls.

in case got rejected? we have duplication covered. normally such situation (file has already been processed) should not happen. so either we doing it wrong or orchestration around here with all these libraries and services doing their own thing is incorrect.

is this just a wild thought or has some value? if no other way - i'm ok to participate in chain of approvals required

Reply:
I thought about acquiring lease straight away but my concern is about performance - if several pods will be acquiring - releasing the lease of the file this can make performance impact, bigger than querying the db